File: leetcode/company-wise/microsoft/0025_reverse_nodes_in_k_group.py
# LC 25. Reverse Nodes in k-Group | Hard | Microsoft
class ListNode:
    def __init__(self, val=0, next=None): self.val = val; self.next = next

class Solution:
    def reverseKGroup(self, head, k):
        count, node = 0, head
        while node and count < k: node = node.next; count += 1
        if count < k: return head
        prev, curr = None, head
        for _ in range(k): nxt = curr.next; curr.next = prev; prev = curr; curr = nxt
        head.next = self.reverseKGroup(curr, k)
        return prev

File: leetcode/company-wise/microsoft/0025_reverse_nodes_in_k_group.cpp
// LC 25. Reverse Nodes in k-Group | Hard | Microsoft
struct ListNode { int val; ListNode* next; ListNode(int x) : val(x), next(nullptr) {} };

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* node = head; int cnt = 0;
        while (node && cnt < k) { node = node->next; cnt++; }
        if (cnt < k) return head;
        ListNode *prev = nullptr, *curr = head;
        for (int i = 0; i < k; i++) { ListNode* nxt = curr->next; curr->next = prev; prev = curr; curr = nxt; }
        head->next = reverseKGroup(curr, k);
        return prev;
    }
};
